File: code/739.每日温度.java
/*
 * @lc app=leetcode.cn id=739 lang=java
 *
 * [739] 每日温度
 *
 * https://leetcode-cn.com/problems/daily-temperatures/description/
 *
 * algorithms
 * Medium (64.46%)
 * Likes:    553
 * Dislikes: 0
 * Total Accepted:    111.6K
 * Total Submissions: 171.4K
 * Testcase Example:  '[73,74,75,71,69,72,76,73]'
 *
 * 请根据每日 气温 列表，重新生成一个列表。对应位置的输出为：要想观测到更高的气温，至少需要等待的天数。如果气温在这之后都不会升高，请在该位置用 0
 * 来代替。
 * 
 * 例如，给定一个列表 temperatures = [73, 74, 75, 71, 69, 72, 76, 73]，你的输出应该是 [1, 1, 4,
 * 2, 1, 1, 0, 0]。
 * 
 * 提示：气温 列表长度的范围是 [1, 30000]。每个气温的值的均为华氏度，都是在 [30, 100] 范围内的整数。
 * 
 */

// @lc code=start
class Solution {
    public int[] dailyTemperatures(int[] T) {
        int[] ans = new int[T.length];
        for (int i = T.length - 2; i >= 0; i--) {
            int curr = i + 1;
            while (T[i] >= T[curr]) {
                if (ans[curr] > 0) {
                    curr += ans[curr];
                } else {
                    curr = i;
                    break;
                }
            }

            ans[i] = curr - i;
        }
        return ans;
    }
}
// class Solution {
//     public int[] dailyTemperatures(int[] T) {
//         int[] ans = new int[T.length];
//         Stack<Integer> stack = new Stack<>();

//         for (int i = 0; i < T.length; i++) {
//             while (!stack.isEmpty() &&  T[i] > T[stack.peek()]) {
//                 int preIdx = stack.pop();
//                 ans[preIdx] = i - preIdx;
//             }
//             stack.push(i);
//         }

//         return ans;
//     }
// }
// @lc code=end

File: code/739.每日温度.go
/*
 * @lc app=leetcode.cn id=739 lang=golang
 *
 * [739] 每日温度
 *
 * https://leetcode-cn.com/problems/daily-temperatures/description/
 *
 * algorithms
 * Medium (67.85%)
 * Likes:    886
 * Dislikes: 0
 * Total Accepted:    211.8K
 * Total Submissions: 311.6K
 * Testcase Example:  '[73,74,75,71,69,72,76,73]'
 *
 * 请根据每日 气温 列表 temperatures ，请计算在每一天需要等几天才会有更高的温度。如果气温在这之后都不会升高，请在该位置用 0 来代替。
 *
 * 示例 1:
 *
 *
 * 输入: temperatures = [73,74,75,71,69,72,76,73]
 * 输出: [1,1,4,2,1,1,0,0]
 *
 *
 * 示例 2:
 *
 *
 * 输入: temperatures = [30,40,50,60]
 * 输出: [1,1,1,0]
 *
 *
 * 示例 3:
 *
 *
 * 输入: temperatures = [30,60,90]
 * 输出: [1,1,0]
 *
 *
 *
 * 提示：
 *
 *
 * 1
 * 30
 *
 *
 */

// @lc code=start
func dailyTemperatures(temperatures []int) []int {
	stack := make([]int, 0)
	res := make([]int, len(temperatures))
	for i, t := range temperatures {
		lastIdx := len(stack) - 1
		for len(stack) > 0 && t > temperatures[stack[lastIdx]] {
			j := stack[lastIdx]
			res[j] = i - j
			stack = stack[:lastIdx]
			lastIdx--
		}
		stack = append(stack, i)
	}
	return res
}

// @lc code=end

